alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportGetRemoteIndexesAction.kt

```diff
@@ -0,0 +1,192 @@
+import org.opensearch.alerting.action.GetRemoteIndexesResponse
+import org.opensearch.alerting.action.GetRemoteIndexesResponse.ClusterIndexes
+import org.opensearch.alerting.action.GetRemoteIndexesResponse.ClusterIndexes.ClusterIndex
+import org.opensearch.alerting.opensearchapi.suspendUntil
+import org.opensearch.alerting.settings.AlertingSettings
+import org.opensearch.alerting.settings.AlertingSettings.Companion.REMOTE_MONITORING_ENABLED
+import org.opensearch.alerting.util.AlertingException
+import org.opensearch.alerting.util.CrossClusterMonitorUtils
+import org.opensearch.client.Client
+import org.opensearch.cluster.service.ClusterService
+import org.opensearch.common.inject.Inject
+import org.opensearch.common.settings.Settings
+import org.opensearch.core.action.ActionListener
+import org.opensearch.core.rest.RestStatus
+import org.opensearch.core.xcontent.NamedXContentRegistry
+import org.opensearch.tasks.Task
+import org.opensearch.transport.TransportService
+import java.time.Duration
+import java.time.Instant
+
+private val log = LogManager.getLogger(TransportGetRemoteIndexesAction::class.java)
+private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO)
+
+class TransportGetRemoteIndexesAction @Inject constructor(
+    val transportService: TransportService,
+    val client: Client,
+    actionFilters: ActionFilters,
+    val xContentRegistry: NamedXContentRegistry,
+    val clusterService: ClusterService,
+    settings: Settings,
+) : HandledTransportAction<GetRemoteIndexesRequest, GetRemoteIndexesResponse>(
+    GetRemoteIndexesAction.NAME,
+    transportService,
+    actionFilters,
+    ::GetRemoteIndexesRequest
+),
+    SecureTransportAction {
+
+    @Volatile override var filterByEnabled = AlertingSettings.FILTER_BY_BACKEND_ROLES.get(settings)
+
+    @Volatile private var remoteMonitoringEnabled = REMOTE_MONITORING_ENABLED.get(settings)
+
+    init {
+        clusterService.clusterSettings.addSettingsUpdateConsumer(REMOTE_MONITORING_ENABLED) { remoteMonitoringEnabled = it }
+        listenFilterBySettingChange(clusterService)
+    }
+
+    override fun doExecute(
+        task: Task,
+        request: GetRemoteIndexesRequest,
+        actionListener: ActionListener<GetRemoteIndexesResponse>
+    ) {
+        log.debug("Remote monitoring enabled: {}", remoteMonitoringEnabled)
+        if (!remoteMonitoringEnabled) {
+            actionListener.onFailure(
+                AlertingException.wrap(
+                    OpenSearchStatusException("Remote monitoring is not enabled.", RestStatus.FORBIDDEN)
+                )
+            )
+            return
+        }
+
+        val user = readUserFromThreadContext(client)
+        if (!validateUserBackendRoles(user, actionListener)) return
+
+        client.threadPool().threadContext.stashContext().use {
+            scope.launch {
+                val singleThreadContext = newSingleThreadContext("GetRemoteIndexesActionThread")
+                withContext(singleThreadContext) {
```

Comment:
Ah interesting. It looks like on this line ISM checks whether the user info is in the context, and injects it if not. Craig and I discussed doing something similar last week, but landed on the single thread implementation since we want to ensure the actions performed by the API take place in a specific sequence. Do you have any insight into the pros and cons of ISM's approach?
https://github.com/opensearch-project/index-management/blob/ba24f862a4b59bdac5a4def76033766346e9eea8/src/main/kotlin/org/opensearch/indexmanagement/opensearchapi/OpenSearchExtensions.kt#L300

Reply:
I think using `ThreadContextElement` is more natural to kotlin coroutine, since it's provided directly by them.
Basically what it does is to attach this element to the coroutine that will execute your code.

While force the user info in one thread, I feel it's possible that this thread may be used to execute other coroutine? then it may cause some problem.